Arabic tokeniser support

AhmedYounes94 opened the pull request:
1- Added all the changes required for Arabic Stanford tokenizer(Similar to Qiwei's work) 
2- Added Arabic stop words 
3- Changed the list of punctuation to include the Arabic one 
4- Added some comments to the code that might need verification 
5- I updated the punctuation file twice to recognize more punctuations 
6- The project now is capable of tokenizing Arabic text, removing stop words (with some limitations), removing punctuation 
7- The limitation in stop words is that the pipeline somewhere apply normalization to the text such as remove hamza from آحمد
like in احمد which effects the recognition of the stop word but I added them normalized too just in case

AhmedYounes94 committed:
1- Changed the version of the dependency parser to 1.18.0 because installing that project by applying mvn install showed this version and without that line the project will not be able to find 1.17.1 version
2- I added a comment in IncrementalSurprisingPhraseAnalysis class saying regarding the punctuation pattern, i am not sure if i need to add anything in this bit as Qiwei mentioned that the pattern he is using is Universal and it should include Arabic maybe you could confirm that for me cause this line might be the cause of some punctuation problems
3- Added Arabic stop words to TokenFilterRelevanceStopwords class, I chose the list in this https://github.com/mohataher/arabic-stop-words repository as it includes all other stop words such as Arabic NLTK stop words and it is the biggest list
4- I added the Arabic option to the pipeline so the user can choose to remove Arabic stop words ConfigHandlerRemoveStopwords

AhmedYounes94 committed:
1- Added Some Arabic Punctuation ،؛«»؟

AhmedYounes94 committed:
1- Add Arabic option in the configurehandler tokeniser
2- Add Arabic Stanford tokeniser dependency with some comments
3- updated the punctuation file to recognise ،
4- Update Arabic stop words list change it to Arabic NLTK stop word list
5- Add TokeniserArabicStanford Class for Arabic tokenisation
6- Create an example for Arabic Tokenisation and stop words removal

andehr commented on pom.xml:
```diff
@@ -9,12 +9,7 @@
 
     <packaging>jar</packaging>
 
-    <parent>
-        <version>1.0.6</version>
-        <groupId>uk.ac.susx.tag</groupId>
-        <artifactId>tag-dist</artifactId>
-        <relativePath>../tag-dist</relativePath>
-    </parent>
```
This is required, why is it removed?

AhmedYounes94 replied:
I removed it because it is mentioned in the read me file

andehr replied:
Ah, that's just to get things running locally without having access to the `tag-dist` repo, but those lines of code should remain in this git repo - they are used for deploying the code.

AhmedYounes94 replied:
Okay added them back

AhmedYounes94 committed:
This is Tha changes i made to create the 4 serializable
 files pipeline, count, inc-cout and samples

1- I changed the ClusteredProcessedInstance and ProcessedInstance class to serializable to be able to serialize it
2- I also changed the code in the ClusterFeatureAnalysis.java  code, This is where i created the four serialized files
3- This is the only thing that need verification
4- The file still throw an error GC overhead limit exceeded
even when i reduced the size of the data to 1000 and it is worth noting that the errors are throw by the stanford code not our code the GC error is somehow thrown from there code (i even reduced the size to 1000 articles and divided them into sentences of 15 words and still not working it says Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
and this error show up after i reduced it into sentences)

andehr commented on src/main/java/uk/ac/susx/tag/classificationframework/datastructures/ProcessedInstance.java:
```diff
@@ -50,7 +51,7 @@
  * Time: 15:39
  */
-public class ProcessedInstance {
+public class ProcessedInstance implements Serializable {
```
Do these classes absolutely need to be java serializable? I'm very reticent to change `ProcessedInstance` or `ClusteredProcessedInstance` since they are used across very many other code repositories.

We currently already serialise instances of documents using JSON, through the Gson library. See for example the [save function in the ModalState class](https://github.com/andehr/tag-classification-framework/blob/aece2465f5fe79d00eb866f51975190bdb56ac65/src/main/java/uk/ac/susx/tag/classificationframework/datastructures/ModelState.java#L178-L225) which saves the inner Instance objects (from trainingDocuments).

AhmedYounes94 replied:
I mentioned in the comments that i needed to that in order to be able to serialize them 
This task was the hardest cause Qiwei doesnt remember so i had to do it this way after a long discussion with jack trying to figure out the right stuff that i need to serialize and as i was doing that i couldnt serialize these classes so i had to do that in order to be able to create the ser file

andehr commented on src/main/java/uk/ac/susx/tag/classificationframework/featureextraction/pipelines/FeatureExtractionPipeline.java:
```diff
@@ -42,6 +42,7 @@
 import uk.ac.susx.tag.classificationframework.featureextraction.inference.featureselection.FeatureSelector;
 import uk.ac.susx.tag.classificationframework.featureextraction.normalisation.TokenNormaliser;
 import uk.ac.susx.tag.classificationframework.featureextraction.tokenisation.Tokeniser;
+import weka.core.Stopwords;
```
Is this import necessary?

AhmedYounes94 replied:
i dont think so i think this one was from Qiwei's work cause it was there when i started i didnt need to import anything for stop words i just added the list

AhmedYounes94 committed:
added parent dependency

andehr commented on src/main/java/uk/ac/susx/tag/classificationframework/clusters/clusteranalysis/ClusterFeatureAnalysis.java:
```diff
@@ -555,6 +558,7 @@ public static void main(String[] args) throws IOException, ClassNotFoundExceptio
                 "algorithms"
         );
 
+        // In the Pipeline i chose basic because Arabic Stanford crashes when gets a whole paragraph instead of sentence it throws this error gc overhead limit exceeded after taking long time
```
Is the basic tokeniser used for the ar-pipeline.ser that gets saved for use with the surprising phrase detector? 

Was the wikipedia background data tokenised using stanford? 

We should ensure that the surprising phrase detector uses the same tokenisation that was used on the background data, otherwise the output will be weird.

Does the basic tokeniser do anything sensible with arabic text? It's just a custom emoticon and url aware simple regex.

AhmedYounes94 replied:
Before anything this bit of work was a mystery i was basically guessing the process no one remembers it and it is probably has been done wrong 
1- Yes i saved the basic tokeniser cause Stanford  keep making an error i think mentioned in the comments 
2- I don't think i tokenised it by Stanford 
3- Yeah i think they are both the same i ignored Stanford in this step cause it was overflowing the memory i have all the errors recorded in my machine so i can reproduce it (maybe when you are back we can talk about the steps that needs to be done in this last phase of so i can make it correctly)
4- Surprisingly the basic tokeniser works fine but of course it is limited with spaces and commas and so on but Arabic is more complicated than that so it works fine as long as everything is written in proper spacing but that is not always the case with Arabic

AhmedYounes94 replied:
I tested the tokeniser and it works fine in the classifier all the options are added but the surprising phrase is not working. I think this is where we should start from.

AhmedYounes94 replied:
<img width="1168" alt="Screenshot 2020-10-29 at 13 56 32" src="https://user-images.githubusercontent.com/68699166/97583218-97863d00-19ee-11eb-9cb7-74601a8508f8.png">
This is a snap shot of the error i am not sure if that helpful but i guess we can talk about it next week when you are back.

AhmedYounes94 committed:
Cleaning ClusteredProcessedInstance class by removing serializable
Cleaning ProcessedInstance class by removing serializable

AhmedYounes94 committed:
Add parent dependency
Clean Arabic stanford dependency from <exclusions> dependencies
Arabic apache commons dependency for reading csv file

AhmedYounes94 committed:
Add new stop word list for more generic performance this list include 750 element. After the addition of this list the tokeniser works perfectly but still the sruprising phrase is not as good as it should be.

AhmedYounes94 committed:
This file is to test the tokenization
1- Create main
2- Build the pipeline
3- Read the Arabic file that contains the text to be tokenized
4- Print the output

AhmedYounes94 committed:
This file is to generate 4 serializable files for sruprising phrases

1- add imports for reading csv files
2- add print statements in saveNewBackgroundCounter function to prompt the user cause it takes long
3- add read csv function
4- add save corpus function to save the articles
5- add save pipeline function to save the pipeline
6- add count_size function for testing
7- add print instance function for testing
8- inside the main read the article file and save it as serializable
9- apply some testing to validate the generated file
10 - do the same thing for pipeline and the rest of the 4 files

AhmedYounes94 committed:
Change classificationframework version to 7.21.4

AhmedYounes94 committed:
Stopwords bug
1- Changed the stopwords list
2- add language variable to handle the static behaviour
3- add getLanguage function with default english
4- change the stopwords retrieval process by accessing getting the language

AhmedYounes94 committed:
Reserialised the pipeline
1- changed getStopwords_dict("en") to stop the errors
2- deserialize the pipeline to see if it still has the lang key

chrisji commented on pom.xml:
```diff
@@ -78,9 +79,12 @@
             <groupId>uk.ac.susx.tag</groupId>
             <artifactId>dependencyparser</artifactId>
-            <version>1.17.1</version>
+            <version>1.18.0</version>
```
- [ ] Simon to confirm why versions have been reverted before.

chrisji commented on pom.xml:
```diff
@@ -161,6 +165,25 @@
+        301 under executions tag there is a plugin defined to load the reflection file or to deploy some script regarding the reflections and this plugin is bound to
+        the compile phase with execute goal and this goal will execute the script configured in the configuration tag and without executing that script the reflections will
+        return null and there will be null reference error (this script is defining and initializing a reflection object that takes as argument a pipeline config11ehandler)-->
+<!--         Note that once you have added the dependency you need to reimport the project i don't really know how to set that automatically-->
```
- [ ] Believe this is just referencing the need for a re-import of dependencies, remove commenting.

chrisji commented on src/main/java/uk/ac/susx/tag/classificationframework/clusters/ClusteredProcessedInstance.java:
```diff
@@ -2,6 +2,8 @@
 
 import uk.ac.susx.tag.classificationframework.datastructures.ProcessedInstance;
 
+import java.io.Serializable;
```
- [ ] remove import

chrisji commented on src/main/java/uk/ac/susx/tag/classificationframework/clusters/clusteranalysis/ClusterFeatureAnalysis.java:
```diff
@@ -366,7 +367,8 @@ public int getNumOfClusters() {
                 numFeatures, numPhrasesPerFeature, pipeline,
                 OrderingMethod.LIKELIHOOD_IN_CLUSTER_OVER_PRIOR, FeatureType.WORD,
-                0.3, 4, 5, 7, 15, TokenFilterRelevanceStopwords.getStopwords(), 1, 6);
+//                Here i changed a static behaviour of getStopwords
```
- [ ] Remove comment

chrisji commented on src/main/java/uk/ac/susx/tag/classificationframework/clusters/clusteranalysis/ClusterFeatureAnalysis.java:
```diff
@@ -539,9 +542,95 @@ public static FeatureBasedCounts loadBackgroundCounter(File inputFile) throws IO
         }
     }
 
+    public static List<Instance> readCsv(String inpath) {
```
- [ ] Remove from PR, is this in the shared ClusterFeatureAnalysis file?

chrisji commented on src/main/java/uk/ac/susx/tag/classificationframework/clusters/clusteranalysis/IncrementalSurprisingPhraseAnalysis.java:
```diff
@@ -36,6 +36,8 @@
     private static final double featureSmoothing = 0.1;
 //     private static final Pattern punct = Pattern.compile("\\p{Punct}+");
     // allow for Chinese puncutation used in cluster analysis
+   /* Ahmed Younes: I am not sure if  i need to add something to this pattern in order to allow Arabic punctuation but as i understood from Qiwei he is using
```
Remove comment

chrisji commented on src/main/java/uk/ac/susx/tag/classificationframework/datastructures/ProcessedInstance.java:
```diff
@@ -23,6 +23,7 @@
 import it.unimi.dsi.fastutil.ints.Int2DoubleMap;
 import it.unimi.dsi.fastutil.ints.Int2DoubleOpenHashMap;
 
+import java.io.Serializable;
```
Remove import

chrisji commented on src/main/java/uk/ac/susx/tag/classificationframework/featureextraction/filtering/TokenFilterPunctuation.java:
```diff
@@ -50,11 +50,17 @@ public TokenFilterPunctuation(boolean filterExclamationAndQuestionMarks){
     }
 
     private Pattern getPunctuationPattern() {
+       /* Ahmed Younes: This is a conditional operator where it could be stated as if a then b otherwise c
```
Clean up

chrisji commented on src/main/java/uk/ac/susx/tag/classificationframework/featureextraction/filtering/TokenFilterRelevanceStopwords.java:
```diff
@@ -1980,10 +1981,32 @@
                 " ",
                 ""));
+        /* here are the Arabic stop words it is added to the dictionary to be automatically assigned once it has been chosen
+         it is a collection of stop words from Arabic NLTK stop words i  also added normalized version cause the stanford tokeniser normalize the text and remove hamza */
```
- [ ]Clean comment, still keep where they were obtained from.

chrisji commented:
See #20
